### DISCUSSION TITLE
More clip play modes

### DISCUSSION THREAD

**Discussion by gramster** [SynthstromAudible/DelugeFirmware#608] [Ideas]
I just got my Deluge yesterday and have barely played with it, but a big reason for getting it was the fact that the code is open source and so it can improve with time. Perhaps at some point I can contribute; been years since I did C++ but I did a *lot* in the 90s.

I do have a Polyend Play, which I think is fantastic, and I would love to see some of the novel features from that make their way to the Deluge (ideally one day the Polyend would be made redundant by the Deluge). One of the first obvious gaps is the clip-play modes; as far as I can see the Deluge has forward, reverse, and ping-pong, while the Polyend has 35, many of which I have yet to figure out what they're actually doing but they sure are fun and interesting. It would be great to see more come to the Deluge, and in particular for it to be easier to add more to the Deluge. I'm a big fan of having lightweight programmable engines that are extensible through config files and this seems like it could be an area where that could be used to good effect, although I probably need to think that through a bit more and perhaps I could come up with a concrete proposal for some declarative DSL for this.

I did do a brief search through the code for 'ping pong' to try to see how things are implemented currently. I need to dig deeper there although I did see a couple of things that suggested to me this might not be a simple change. In particular, I noticed a Boolean flag 'currentlyPlayingReversed' which makes some sense if all you have is forward/reverse or ping-pong, but probably makes no sense if you have some more creative modes. Intuitively to me, it would make sense to generate an array of step indices whenever a play mode is toggled or a clip length is changed, and then use that plus a current offset that cycles through the array as a general mechanism that could incorporate many play modes with a single mechanism (at the cost of a bit of memory, but maybe these arrays could be shared and be run-length encoded, etc, to reduce that). Admittedly on the Polyend Play all this is made easier by steps being pretty self-contained (many things like sample, MIDI channel, etc, are per step rather than per-track). Anyway, wanted to kick off this discussion and hope to learn a lot more going forward (see what I did there ;-)).

**Comment by saynotohugs** [comment]
Hey Gramster

This is really interesting :) I agree, to have some more play modes like on the polyend would definately be fantastic :)

**Comment by gramster** [comment]
The Polyend Play modes are documented now; page 46-47 of the manual here https://polyend.com/wp-content/uploads/2023/11/Play_Plus_Manual_Rev1.pdf

**Reply by sapphire-arches** [reply]
In general the playback methods seem to be pretty tied to the idea that patterns are 16/32 steps long, which might not play super nicely with the Deluge's very flexible track system -- individual note rows can have their own lengths, not constrained to a multiple of 16 steps.

That's not to say these couldn't be adapted to the Deluge sequencer! But I think we should try to avoid tying such effects to particular numbers of beats/bars -- maybe some configuration of the "break point" for each sequencer effect, as it seems like each effect has 1 step it treats differently.

The largest technical challenge is that the Deluge sequencer currently schedules the next note once the current note is executed, and the logic for doing that is pretty intricate. It would likely need to be refactored and cleaned up before it could be made sufficiently flexible to handle this sort of step jumping.
